File: 3Sum.swift
class Solution {
    // Notes
    // "a" will always be negative -a
    // b + c = -a
    // c = -(a + b)
    // "c" will always be unique because a pair of unique [a, b] will result in a unique "c" value
    """
    Time complexity: O(n2)
    Space complexity: O(n)
    
    n is the number of elements in the input array nums.
    The outer forEach iterates n times, and for each iteration, pairSumSorted runs in O(n) with its two-pointer approach, resulting in O(n^2).
    The space complexity is O(n) for the sorted array and the tripples output storage.
    """
    func threeSum(_ nums: [Int]) -> [[Int]] {
        var tripples = [[Int]]()
        let sorted = nums.sorted()
        sorted.enumerated().forEach { index, value in
            guard value <= 0 else { return }

            if index > 0, index + 1 < sorted.count, value == sorted[index - 1] {
                return
            }

            let pairs = pairSumSorted(nums: sorted, startIndex: index + 1, target: -value) 
            tripples += pairs.map { [value] + $0 }
        }

        return tripples
    }

    func pairSumSorted(nums: [Int], startIndex: Int, target: Int) -> [[Int]] {
        var leftIndex = startIndex
        var rightIndex = nums.count - 1
        var pairs = [[Int]]()

        while leftIndex < rightIndex {
            let sum = nums[leftIndex] + nums[rightIndex]

            if sum == target {
                pairs.append([nums[leftIndex], nums[rightIndex]])
                leftIndex += 1
                // To avoid duplicate [b, c] pairs, skip "b"
                while leftIndex < rightIndex && (nums[leftIndex] == nums[leftIndex - 1]) {
                    leftIndex += 1
                }
            } else if sum < target {
                leftIndex += 1
            } else {
                rightIndex -= 1
            }
        }

        return pairs
    }
}

File: 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        triplets = []
        nums.sort()
        for i in range(len(nums)):
            #should be always 0 or nagative number
            if nums[i] > 0:
                break
            #Avoid duplications    
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            # Target: b + c = -a    
            pairs = self.sortedPairs(nums, i + 1, -nums[i])
            for pair in pairs:
                triplets.append([nums[i]] + pair)
        return triplets

    def sortedPairs(self, nums: List[int], startIndex: int, target: int) -> List[List[int]]:
        pairs = []
        left = startIndex
        right = len(nums) - 1

        while left < right:
            sum = nums[left] + nums[right]
            if sum == target:
                pairs.append([nums[left], nums[right]])
                left += 1
                while left < right and nums[left] == nums[left - 1]:
                    left += 1
            elif sum < target:
                left += 1
            else:
                right -= 1     
        return pairs       
        
